migration.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/application-research/estuary/util"
+	"gorm.io/gorm"
+)
+
+type Migration struct {
+	Name        string `gorm:"primarykey"`
+	StartedAt   time.Time
+	CompletedAt time.Time
+}
+
+// ordered list of migrations. Data migrations are normally executed in the background so a migration function must
+// not interfere with the running of the rest of estuary and should be stateless, interruptable and resumable. A
+// migration may be marked as required in which case estuary will not start running until the migration has completed.
+// A large data migration may be implemented in stages across multiple versions of estuary. The first stage would
+// migrate the data in the background, retaining compatibility with existing queries. Once this is complete a later
+// version of estuary may simplify code by removing the compatibility and updating the migration to be required. This
+// will have no effect on systems that have already migrated but will allow older systems to be updated safely.
+var migrations = []struct {
+	name string
+	fn   func(context.Context, *gorm.DB) error
+
+	// when true, estuary will not start until the migration has completed
+	required bool
+}{
+
+	{
+		// Migration from cid to multihash for the objects table
+		name: "object-cid-to-hash",
+		fn: func(ctx context.Context, db *gorm.DB) error {
+			var objs []*Object
+			tx := db.Where("hash IS NULL").FindInBatches(&objs, 500, func(tx *gorm.DB, batch int) error {
+				// Make sure migration is interruptable
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				default:
+				}
+
+				for _, o := range objs {
+					o.Hash = util.DbHashFromDbCID(&o.Cid)
+				}
+
+				tx.Save(&objs)
```

Comment:
the `Save` method in gorm gives me trouble somewhat often. I've switched all my calls of it to just updating the specific field I want to update.

Reply:
Also, is calling this on the DB instance from the closure the right thing? I'm not really sure what the implications are here